---
title: "OmicsMLRepoR - Quickstart"
author: "Sehyun Oh, Kaelyn Long"
date: "`r format(Sys.time(), '%B %d, %Y')`"
vignette: >
  %\VignetteIndexEntry{Quickstart}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
output:
  BiocStyle::html_document:
    number_sections: yes
    toc: yes
    toc_depth: 4
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(comment = "#>", 
                      collapse = TRUE, 
                      message = FALSE, 
                      warning = FALSE)
```

# Introduction

Our OmicsMLRepo project aims to improve the AI/ML-readiness of Omics
datasets available through Bioconductor. One of the main activities
under this project is metadata harmonization and standardization (i.e.,
incorporating ontology).

Currently, we released the harmonized version of metadata for two
Bioconductor data packages -
[curatedMetagenomicData](https://www.bioconductor.org/packages/release/data/experiment/html/curatedMetagenomicData.html)
containing human microbiome data and
[cBioPortalData](https://www.bioconductor.org/packages/release/bioc/html/cBioPortalData.html)
package on cancer genomics data. **OmicsMLRepoR** is a software package
allowing users to easily access the harmonized metadata and to leverage
ontology in metadata search.

`OmicsMLRepoR` provides the two major functions:     
1. Access to harmonized metadata    
2. Browse data using ontology    


## Load package

```{r eval=FALSE}
if (!require("devtools"))
    install.packages("devtools")
devtools::install_github("shbrief/OmicsMLRepoR")
```

```{r}
suppressPackageStartupMessages({
    library(OmicsMLRepoR)
    library(dplyr)
    library(curatedMetagenomicData)
})
```

## Load the metadata

You can download the harmonized version of metadata using the
`getMetadata` function. Currently, two options are available - `cMD` and
`cBioPortalData`.

```{r}
cmd <- getMetadata("cMD")
cmd
```

```{r}
cbio <- getMetadata("cBioPortal")
cbio
```

# Access metadata

## Search using ontology

[Todo] Update `searchMetadata` to convert collapsed metadata into the
long version to detect the descendants.

```{r}
# res1 <- searchMetadata("cancer", cmd)
# res2 <- searchMetadata("cancer", cmd, "target_condition")
```

## `tree_filter`

```{r}
nrow(cmd |> tree_filter(disease, c("Healthy")))
nrow(cmd |> tree_filter(disease, c("Colorectal Carcinoma")))


nrow(cmd |> tree_filter(disease, c("Healthy|Colorectal Carcinoma")))
nrow(cmd |> tree_filter(disease, c("Healthy&Colorectal Carcinoma"), "cMD"))
nrow(cmd |> tree_filter(disease, c("Colorectal Carcinoma"), "cMD"))
```

## Robust search using ontology

Metadata searching in the original `curatedMetagenomicData` was not
robust. For example, it's sensitive to the cases and doesn't recognize
abbreviations.

### Not case-sensitive
```{r}
nrow(cmd |> tree_filter(disease, c("healthy", "colorectal carcinoma")))
nrow(cmd |> tree_filter(disease, c("Healthy", "Colorectal Carcinoma")))
```

### Include synonyms
```{r}
syn_res1 <- cmd |> tree_filter(disease, "CRC")
syn_res2 <- cmd |> tree_filter(disease, "Colorectal Cancer")
syn_res3 <- cmd |> tree_filter(disease, "Colorectal Carcinoma")

nrow(syn_res1)
nrow(syn_res2)
nrow(syn_res3)

unique(syn_res1$disease)
unique(syn_res2$disease)
unique(syn_res3$disease)
```

### Search descendants in ontology tree
```{r}
onto_res <- cmd |> tree_filter(disease, "Intestinal Disorder")
unique(onto_res$disease)
```

## Collapse/Expand metadata

Some metadata columns (e.g., `biomarker`) contain multiple, similar
attributes separated with a specific delimiter (i.e., `<;>`). Our
harmonization use this structure because they are related information
often looked up together.

```{r}
cmd_biomarker <- cmd %>% 
    filter(!is.na(biomarker)) %>% 
    select(curation_id, biomarker)
wtb <- getWideMetaTb(cmd_biomarker, "biomarker")
```

```{r}
ltb <- getLongMetaTb(cmd, targetCols = "target_condition")
```

# Download

Harmonized metadata is used to extract the searched data through their
original packages - curatedMetagenomicData and cBioPortalData.

## Query relative abundance

## Query RNA sequencing

## Extract subset of samples from a single study

## Extract multiple studies

## Extract subset of samples from multiple studies

# Obtain unharmonized metadata
